source/common/network/listener_impl.cc

```diff
@@ -27,7 +27,11 @@ void ListenerImpl::listenCallback(evconnlistener*, evutil_socket_t fd, sockaddr*
                                   int remote_addr_len, void* arg) {
   ListenerImpl* listener = static_cast<ListenerImpl*>(arg);
 
-  Address::InstanceConstSharedPtr final_local_address = listener->socket_.localAddress();
+  // Get the local address from the new socket.  This is different from the listener address, if
+  // the listener is listening on the all hosts address (e.g., 0.0.0.0 for IPv4).
+  Address::InstanceConstSharedPtr local_address = Address::addressFromFd(fd);
```

Comment:
Alright, having read through the description of what you are trying to accomplish, this change sounds reasonable to me, so let's go ahead and add tests and get it shipped. As you point out already, I would recommend adding some type of interface that can perform the various socket option lookups, etc. that we can mock in tests. I realize this code is not very well tested today in that regard, but it "works" and if we are going to change it we should improve the test coverage.